### PR TITLE
HBASE-25596: Fix NPE and avoid permanent unreplicated data due to EOF

### PR DESCRIPTION
**Pull request by sandeepvinayak**
Please find the details in jira:
https://issues.apache.org/jira/browse/HBASE-25596

### CODE REVIEW

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java [hunk 1]
```diff
@@ -241,11 +305,22 @@ private void handleEofException(Exception e) {
...
           lastReadPath = logQueue.remove();
           lastReadPosition = 0;
+
+          // If it was on last log in the recovered queue,
+          // the stream doesn't have more data, we should stop the reader
+          boolean hasMoreData = !logQueue.isEmpty();
+          // After we removed the WAL from the queue, we should
+          // try shipping the existing batch of entries, we do not want to reset
+          // stream since entry stream doesn't have the correct data at this point
+          isBatchQueuedToBeShipped(entryStream, batch, hasMoreData, true);
```
Comment (sandeepvinayak) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:315]: This is the fix to avoid permanent data loss.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSource.java [hunk 2]
```diff
@@ -237,6 +237,10 @@ public void enqueueLog(Path log) {
     }
   }
 
+  public Map<String, PriorityBlockingQueue<Path>> getQueues() {
```
Comment (apurtell) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSource.java:240]: If this is for tests only, you can make this package scope

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSource.java [hunk 3]
```diff
@@ -840,7 +844,7 @@ public void uncaughtException(final Thread t, final Throwable e) {
 
     // If this is a recovered queue, the queue is already full and the first log
     // normally has a position (unless the RS failed between 2 logs)
-    private long getRecoveredQueueStartPos(long startPosition) {
+    public long getRecoveredQueueStartPos(long startPosition) {
```
Comment (apurtell) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSource.java:847]: Is this change needed? I don't see any changes to a caller of this method in this patch

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java [hunk 4]
```diff
@@ -133,71 +134,126 @@ public ReplicationSourceWALReaderThread(ReplicationSourceManager manager,
...
-                long entrySizeExcludeBulkLoad = getEntrySizeExcludeBulkLoad(entry);
-                batch.addEntry(entry, entrySize);
-                updateBatchStats(batch, entry, entryStream.getPosition(), entrySize);
-                boolean totalBufferTooLarge = acquireBufferQuota(entrySizeExcludeBulkLoad);
-                // Stop if too many entries or too big
-                if (totalBufferTooLarge || batch.getHeapSize() >= replicationBatchSizeCapacity
+    WALEntryBatch batch = null;
+    WALEntryStream entryStream =
+      new WALEntryStream(logQueue, fs, conf, lastReadPosition, metrics);
+    try {
```
Comment (apurtell) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:140]: There is a new outer try block added here...

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java [hunk 5]
```diff
@@ -133,71 +134,126 @@ public ReplicationSourceWALReaderThread(ReplicationSourceManager manager,
...
-        } else {
-          LOG.error("Failed to read stream of replication entries", e);
-          handleEofException(e);
-        }
-        Threads.sleep(sleepForRetries * sleepMultiplier);
-      } catch (InterruptedException e) {
-        LOG.trace("Interrupted while sleeping between WAL reads");
-        Thread.currentThread().interrupt();
       }
+    } catch (IOException e) {
```
Comment (apurtell) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:200]: ... and the catch clauses from the inner try have been moved out to the outer try. That's the idea, right? Or did I misread this

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java [hunk 6]
```diff
@@ -133,71 +134,126 @@ public ReplicationSourceWALReaderThread(ReplicationSourceManager manager,
...
+            } else {
+              LOG.error("Failed to read stream of replication entries", e);
+            }
+            Threads.sleep(sleepForRetries * sleepMultiplier);
           }
-          resetStream(entryStream);
+        } catch (InterruptedException e) {
+          LOG.trace("Interrupted while sleeping between WAL reads");
+          Thread.currentThread().interrupt();
+        } finally {
```
Comment (apurtell) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:196]: ... and the catch clause for the inner try now just ensures 'entryStream' is closed ...

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java [hunk 7]
```diff
@@ -133,71 +134,126 @@ public ReplicationSourceWALReaderThread(ReplicationSourceManager manager,
...
+   * @param batch Batch of entries to ship
+   * @param hasMoreData if the stream has more yet more data to read
+   * @param isEOFException if we have hit the EOF exception before this. For EOF exception,
+   *                      we do not want to reset the stream since entry stream doesn't
+   *                      have correct information.
+   * @return if batch is shipped successfully
+   * @throws InterruptedException throws interrupted exception
+   * @throws IOException throws io exception from stream
+   */
+  private boolean isBatchQueuedToBeShipped(WALEntryStream entryStream, WALEntryBatch batch,
```
Comment (apurtell) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:226]: This method does more than just check for a condition and return a boolean, it also takes action. Giving it a name isFoo is misleading. I guess you could call it updateAndShip or updateAndShipBatch or whatever. This is only called from one place, inside handleEofException, just put this code in there inline?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java [hunk 8]
```diff
@@ -227,10 +283,18 @@ private void resetStream(WALEntryStream stream) throws IOException {
     stream.reset(); // reuse stream
   }
 
-  // if we get an EOF due to a zero-length log, and there are other logs in queue
-  // (highly likely we've closed the current log), we've hit the max retries, and autorecovery is
-  // enabled, then dump the log
-  private void handleEofException(Exception e) {
+  /**
+   * This is to handle the EOFException from the WAL entry stream. EOFException should
```
Comment (apurtell) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:287]: Good comment

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java [hunk 9]
```diff
@@ -405,7 +481,7 @@ public long getLastReadPosition() {
    * Holds a batch of WAL entries to replicate, along with some statistics
    *
    */
-  static class WALEntryBatch {
+  final static class WALEntryBatch {
```
Comment (apurtell) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:484]: Not directly related but this is a good change

File: hbase-server/src/test/java/org/apache/hadoop/hbase/replication/TestReplicationSource.java [hunk 10]
```diff
@@ -34,6 +34,8 @@
 import static org.mockito.Mockito.when;
 import static org.mockito.internal.verification.VerificationModeFactory.times;
 
+import com.google.common.collect.Lists;
```
Comment (apurtell) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/replication/TestReplicationSource.java:37]: This should come from hbase-thirdparty e.g. org.apache.hbase.thirdparty.com.google.common.collect.Lists

File: hbase-server/src/test/java/org/apache/hadoop/hbase/replication/TestReplicationSource.java [hunk 11]
```diff
@@ -542,6 +556,173 @@ public WALEntryFilter getWALEntryfilter() {
     });
   }
 
+  @Test
```
Comment (apurtell) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/replication/TestReplicationSource.java:559]: Nice test

File: hbase-server/src/test/java/org/apache/hadoop/hbase/replication/regionserver/TestWALEntryStream.java [hunk 12]
```diff
@@ -726,8 +727,62 @@ public void testEOFExceptionForRecoveredQueue() throws Exception {
         queue, 0, fs, conf, getDummyFilter(),
         new MetricsSource("1"), (ReplicationSource) source);
     reader.run();
+    assertEquals(0, queue.size());
+  }
+
+  @Test
```
Comment (apurtell) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/replication/regionserver/TestWALEntryStream.java:733]: Nice test

**Review by apurtell — CHANGES_REQUESTED** [review]

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSource.java (hunk 3 above)
Comment (sandeepvinayak) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSource.java:847]: No, I will revert it.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java (hunk 4 above)
Comment (sandeepvinayak) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:140]: yes, I added it here since it can blow up again in the `finally` while `entryStream.close()`

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java (hunk 6 above)
Comment (sandeepvinayak) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:196]: yes

File: hbase-server/src/test/java/org/apache/hadoop/hbase/replication/TestReplicationSource.java (hunk 10 above)
Comment (apurtell) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/replication/TestReplicationSource.java:37]: Never mind, this is branch-1. ignore this comment. I will resolve it.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java (hunk 5 above)
Comment (sandeepvinayak) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:200]: yes, the internal catch should ideally handle the internal try, but since internal `catch` may also blow up since now we are trying to add the data in batch from there. So, we are handling that in the external catch.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java (hunk 7 above)
Comment (sandeepvinayak) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:226]: yes, make sense to rename it, and to answer your second question, it is called from two places, from handleEofException and normal flow.

File: hbase-server/src/test/java/org/apache/hadoop/hbase/replication/TestReplicationSource.java (hunk 10 above)
Comment (sandeepvinayak) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/replication/TestReplicationSource.java:37]: @apurtell I can find guava in thirdparty, thus every place in hbase uses `Lists` directly from com.google.

**Review by apurtell — APPROVED** [review]

File: hbase-server/src/test/java/org/apache/hadoop/hbase/replication/regionserver/TestWALEntryStream.java [hunk 13]
```diff
@@ -726,8 +727,62 @@ public void testEOFExceptionForRecoveredQueue() throws Exception {
...
+    // Override the max retries multiplier to fail fast.
+    conf.setInt("replication.source.maxretriesmultiplier", 1);
+    conf.setBoolean("replication.source.eof.autorecovery", true);
+    // Create a reader thread.
+    ReplicationSourceWALReaderThread reader =
+      new ReplicationSourceWALReaderThread(mockSourceManager, getRecoveredQueueInfo(),
+        queue, 0, fs, conf, getDummyFilter(),
+        new MetricsSource("1"), (ReplicationSource) source);
+    reader.run();
+
```
Comment (xcangCRM) [inline hbase-server/src/test/java/org/apache/hadoop/hbase/replication/regionserver/TestWALEntryStream.java:766]: do you need to do some wait here? How did you drain the queue? I might be missing something here. thanks

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java [hunk 14]
```diff
@@ -133,71 +134,126 @@ public ReplicationSourceWALReaderThread(ReplicationSourceManager manager,
...
+              sleepMultiplier = 1;
             }
+          }
+        } catch (IOException | WALEntryStreamRuntimeException e) { // stream related
+          if (handleEofException(e, entryStream, batch)) {
+            sleepMultiplier = 1;
           } else {
-            Thread.sleep(sleepForRetries);
+            if (sleepMultiplier < maxRetriesMultiplier) {
+              LOG.debug("Failed to read stream of replication entries: " + e);
```
Comment (xcangCRM) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:186]: Do you think it may be valuable to log as info for the failed reason even we are going to retry it? 
In case the final time exception is different from early tries'.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java [hunk 15]
```diff
@@ -133,71 +134,126 @@ public ReplicationSourceWALReaderThread(ReplicationSourceManager manager,
...
+              LOG.error("Failed to read stream of replication entries", e);
+            }
+            Threads.sleep(sleepForRetries * sleepMultiplier);
           }
-          resetStream(entryStream);
+        } catch (InterruptedException e) {
+          LOG.trace("Interrupted while sleeping between WAL reads");
+          Thread.currentThread().interrupt();
+        } finally {
+          entryStream.close();
```
Comment (xcangCRM) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:197]: will we be able to put entryStream into try-with-resource paradigm  above?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java [hunk 16]
```diff
@@ -133,71 +134,126 @@ public ReplicationSourceWALReaderThread(ReplicationSourceManager manager,
...
+   * @param batch Batch of entries to ship
+   * @param hasMoreData if the stream has more yet more data to read
+   * @param isEOFException if we have hit the EOF exception before this. For EOF exception,
+   *                      we do not want to reset the stream since entry stream doesn't
+   *                      have correct information.
+   * @return if batch is shipped successfully
+   * @throws InterruptedException throws interrupted exception
+   * @throws IOException throws io exception from stream
+   */
+  private boolean updateBatchAndAddInShippingQueue(WALEntryStream entryStream, WALEntryBatch batch,
```
Comment (xcangCRM) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:226]: nit:
name is bit long,
how about updateBatchAndShippingQueue

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java [hunk 17]
```diff
@@ -133,71 +134,126 @@ public ReplicationSourceWALReaderThread(ReplicationSourceManager manager,
...
+      boolean hasMoreData, boolean isEOFException) throws InterruptedException, IOException {
+    updateBatch(entryStream, batch, hasMoreData, isEOFException);
+    boolean isDataQueued = false;
+    if (isShippable(batch)) {
+      isDataQueued = true;
+      entryBatchQueue.put(batch);
+      if (!batch.hasMoreEntries()) {
+        // we're done with queue recovery, shut ourselves down
```
Comment (xcangCRM) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:234]: worth adding a debugging log here?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java [hunk 18]
```diff
@@ -511,9 +587,9 @@ public boolean isEmpty() {
       return walEntriesWithSize.isEmpty();
     }
 
-    public void updatePosition(WALEntryStream entryStream) {
-      lastWalPath = entryStream.getCurrentPath();
-      lastWalPosition = entryStream.getPosition();
+    public void updatePosition(Path currentPath, long currentPosition) {
```
Comment (xcangCRM) [inline hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:590]: can you add javadoc on this method to make it clear what's expected for these parameters since it's a public method?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java (hunk 14 above)
Comment (sandeepvinayak) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:186]: @xcangCRM  I haven't changed the log level here, but I think debug makes more sense here since there is an exception involved. What do you think?

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java (hunk 15 above)
Comment (sandeepvinayak) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:197]: initially it was like that way, then entryStream will not be accessible from catch block, that is the reason of removing it from autocloseable resource

File: hbase-server/src/test/java/org/apache/hadoop/hbase/replication/regionserver/TestWALEntryStream.java (hunk 13 above)
Comment (sandeepvinayak) [reply hbase-server/src/test/java/org/apache/hadoop/hbase/replication/regionserver/TestWALEntryStream.java:766]: @xcangCRM i think you are confusing it with `.start()` method which is async call. We are executing `.run()` which is sync call.

**Comment by Apache-HBase** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   3m 52s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 3 new or modified test files.  |
||| _ branch-1 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  11m 17s |  branch-1 passed  |
| +1 :green_heart: |  compile  |   0m 59s |  branch-1 passed with JDK Azul Systems, Inc.-1.8.0_262-b19  |
| +1 :green_heart: |  compile  |   1m  0s |  branch-1 passed with JDK Azul Systems, Inc.-1.7.0_272-b10  |
| +1 :green_heart: |  checkstyle  |   2m 19s |  branch-1 passed  |
| +1 :green_heart: |  shadedjars  |   4m 33s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  7s |  branch-1 passed with JDK Azul Systems, Inc.-1.8.0_262-b19  |
| +1 :green_heart: |  javadoc  |   1m  0s |  branch-1 passed with JDK Azul Systems, Inc.-1.7.0_272-b10  |
| +0 :ok: |  spotbugs  |   4m 13s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   4m  8s |  branch-1 passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 51s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  3s |  the patch passed with JDK Azul Systems, Inc.-1.8.0_262-b19  |
| +1 :green_heart: |  javac  |   1m  3s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  4s |  the patch passed with JDK Azul Systems, Inc.-1.7.0_272-b10  |
| +1 :green_heart: |  javac  |   1m  4s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   2m 14s |  hbase-server: The patch generated 0 new + 0 unchanged - 10 fixed = 0 total (was 10)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   4m 40s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |   6m 48s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2.  |
| +1 :green_heart: |  javadoc  |   0m 57s |  hbase-server-jdkAzulSystems,Inc.-1.8.0_262-b19 with JDK Azul Systems, Inc.-1.8.0_262-b19 generated 0 new + 6 unchanged - 2 fixed = 6 total (was 8)  |
| +1 :green_heart: |  javadoc  |   1m  3s |  hbase-server-jdkAzulSystems,Inc.-1.7.0_272-b10 with JDK Azul Systems, Inc.-1.7.0_272-b10 generated 0 new + 6 unchanged - 2 fixed = 6 total (was 8)  |
| +1 :green_heart: |  findbugs  |   4m 37s |  the patch passed  |
||| _ Other Tests _ |
| -1 :x: |  unit  | 150m 16s |  hbase-server in the patch failed.  |
| +1 :green_heart: |  asflicense  |   0m 29s |  The patch does not generate ASF License warnings.  |
|  |   | 211m 35s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hbase.regionserver.TestSplitTransactionOnCluster |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2975/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2975 |
| JIRA Issue | HBASE-25596 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux f81601d65ef1 4.15.0-101-generic #102-Ubuntu SMP Mon May 11 10:07:26 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-home/workspace/Base-PreCommit-GitHub-PR_PR-2975/out/precommit/personality/provided.sh |
| git revision | branch-1 / f9a9148 |
| Default Java | Azul Systems, Inc.-1.7.0_272-b10 |
| Multi-JDK versions | /usr/lib/jvm/zulu-8-amd64:Azul Systems, Inc.-1.8.0_262-b19 /usr/lib/jvm/zulu-7-amd64:Azul Systems, Inc.-1.7.0_272-b10 |
| unit | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2975/1/artifact/out/patch-unit-hbase-server.txt |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2975/1/testReport/ |
| Max. process+thread count | 3660 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2975/1/console |
| versions | git=1.9.1 maven=3.0.5 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java (hunk 14 above)
Comment (sandeepvinayak) [reply hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/ReplicationSourceWALReaderThread.java:189]: @xcangCRM  I think that is why it is to debug because other on retries it will explode the log file

**Review by xcangCRM — APPROVED** [review]

**Comment by Apache-HBase** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   2m 38s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  hbaseanti  |   0m  0s |  Patch does not have any anti-patterns.  |
| +1 :green_heart: |  @author  |   0m  1s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 3 new or modified test files.  |
||| _ branch-1 Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  10m 48s |  branch-1 passed  |
| +1 :green_heart: |  compile  |   1m  8s |  branch-1 passed with JDK Azul Systems, Inc.-1.8.0_262-b19  |
| +1 :green_heart: |  compile  |   1m  6s |  branch-1 passed with JDK Azul Systems, Inc.-1.7.0_272-b10  |
| +1 :green_heart: |  checkstyle  |   2m 28s |  branch-1 passed  |
| +1 :green_heart: |  shadedjars  |   4m 28s |  branch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  javadoc  |   1m  3s |  branch-1 passed with JDK Azul Systems, Inc.-1.8.0_262-b19  |
| +1 :green_heart: |  javadoc  |   0m 56s |  branch-1 passed with JDK Azul Systems, Inc.-1.7.0_272-b10  |
| +0 :ok: |  spotbugs  |   4m 21s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   4m 18s |  branch-1 passed  |
| -0 :warning: |  patch  |   4m 28s |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   2m 48s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 59s |  the patch passed with JDK Azul Systems, Inc.-1.8.0_262-b19  |
| +1 :green_heart: |  javac  |   0m 59s |  the patch passed  |
| +1 :green_heart: |  compile  |   1m  1s |  the patch passed with JDK Azul Systems, Inc.-1.7.0_272-b10  |
| +1 :green_heart: |  javac  |   1m  1s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   2m 16s |  hbase-server: The patch generated 0 new + 0 unchanged - 10 fixed = 0 total (was 10)  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedjars  |   4m  6s |  patch has no errors when building our shaded downstream artifacts.  |
| +1 :green_heart: |  hadoopcheck  |   5m 56s |  Patch does not cause any errors with Hadoop 2.8.5 2.9.2.  |
| +1 :green_heart: |  javadoc  |   0m 33s |  the patch passed with JDK Azul Systems, Inc.-1.8.0_262-b19  |
| +1 :green_heart: |  javadoc  |   0m 43s |  the patch passed with JDK Azul Systems, Inc.-1.7.0_272-b10  |
| +1 :green_heart: |  findbugs  |   3m  9s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  | 122m 57s |  hbase-server in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 31s |  The patch does not generate ASF License warnings.  |
|  |   | 178m 15s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.41 ServerAPI=1.41 base: https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2975/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hbase/pull/2975 |
| JIRA Issue | HBASE-25596 |
| Optional Tests | dupname asflicense javac javadoc unit spotbugs findbugs shadedjars hadoopcheck hbaseanti checkstyle compile |
| uname | Linux 471c6b614fb2 4.15.0-101-generic #102-Ubuntu SMP Mon May 11 10:07:26 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /home/jenkins/jenkins-home/workspace/Base-PreCommit-GitHub-PR_PR-2975/out/precommit/personality/provided.sh |
| git revision | branch-1 / 9e9301a |
| Default Java | Azul Systems, Inc.-1.7.0_272-b10 |
| Multi-JDK versions | /usr/lib/jvm/zulu-8-amd64:Azul Systems, Inc.-1.8.0_262-b19 /usr/lib/jvm/zulu-7-amd64:Azul Systems, Inc.-1.7.0_272-b10 |
|  Test Results | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2975/3/testReport/ |
| Max. process+thread count | 3693 (vs. ulimit of 10000) |
| modules | C: hbase-server U: hbase-server |
| Console output | https://ci-hadoop.apache.org/job/HBase/job/HBase-PreCommit-GitHub-PR/job/PR-2975/3/console |
| versions | git=1.9.1 maven=3.0.5 findbugs=3.0.1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by apurtell** [comment]
Unless there is a change in review status I will merge this in a couple of hours.
